variadic macros - no argument for ellipsis creates preprocessor parser error

The sample on MSDN for variadic macros will create a parser error.
https://msdn.microsoft.com/en-us/library/ms177415.aspx


excerpt from sample to reproduce the Parser error:
```
#include <stdio.h>
#define EMPTY
#define MACRO(s, ...) printf(s, __VA_ARGS__)

int main() {
    MACRO("hello, world\n");

    MACRO("error\n", EMPTY); // would cause error C2059, except VC++ 
                             // suppresses the trailing comma

}
```

--- a/cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CxxPreprocessor.java
@@ -60,6 +60,7 @@ import static com.sonar.sslr.api.GenericTokenType.IDENTIFIER;
 
 import static org.sonar.cxx.api.CppKeyword.IFDEF;
 import static org.sonar.cxx.api.CppKeyword.IFNDEF;
+import static org.sonar.cxx.api.CppPunctuator.COMMA;
 import static org.sonar.cxx.api.CppPunctuator.LT;
 import static org.sonar.cxx.api.CxxTokenType.NUMBER;
 import static org.sonar.cxx.api.CxxTokenType.PREPROCESSOR;
@@ -869,14 +870,29 @@ public class CxxPreprocessor extends Preprocessor {
             }
           }
 
-          newTokens.add(Token.builder()
-            .setLine(replacement.getLine())
-            .setColumn(replacement.getColumn())
-            .setURI(replacement.getURI())
-            .setValueAndOriginalValue(newValue)
-            .setType(replacement.getType())
-            .setGeneratedCode(true)
-            .build());
+          if (newValue.isEmpty() && "__VA_ARGS__".equals(curr.getValue())) {
+            // the Visual C++ implementation will suppress a trailing comma
+            // if no arguments are passed to the ellipsis
+            for (int n = newTokens.size() - 1; n != 0; n = newTokens.size() - 1) {
+              if (newTokens.get(n).getType() == WS) {
+                newTokens.remove(n);
+              } else if (newTokens.get(n).getType() == COMMA) {
+                newTokens.remove(n);
+                break;
+              } else {
+                break;
+              }
+            }
+          } else {
+            newTokens.add(Token.builder()
+              .setLine(replacement.getLine())
+              .setColumn(replacement.getColumn())
+              .setURI(replacement.getURI())
+              .setValueAndOriginalValue(newValue)
+              .setType(replacement.getType())
+              .setGeneratedCode(true)
+              .build());
+          }
         }
       }
     }

--- a/cxx-squid/src/test/java/org/sonar/cxx/parser/PreprocessorDirectivesTest.java
+++ b/cxx-squid/src/test/java/org/sonar/cxx/parser/PreprocessorDirectivesTest.java
@@ -175,7 +175,7 @@ public class PreprocessorDirectivesTest extends ParserBaseTest {
       + "string s = foo(bar);"))
       .equals("string s = bar + \"x\" ; EOF"));
   }
-
+  
   @Test
   public void variadic_macros() {
     assert (serialize(p.parse(
@@ -198,13 +198,21 @@ public class PreprocessorDirectivesTest extends ParserBaseTest {
             + "eprintf(\"%s:%d: \", input_file, lineno);"))
         .equals("fprintf ( stderr , \"%s:%d: \" , input_file , lineno ) ; EOF"));
 
-    //without whitespace after the parameter list
+    // the Visual C++ implementation will suppress a trailing comma
+    // if no arguments are passed to the ellipsis
+    assert (serialize(p.parse(
+        "#define EMPTY\n"
+            + "#define MACRO(s, ...) printf(s, __VA_ARGS__)\n"
+            + "MACRO(\"error\", EMPTY);"))
+        .equals("printf ( \"error\" ) ; EOF"));
+       
+    // without whitespace after the parameter list
     assert (serialize(p.parse(
         "#define foo(a...);\n"
             + "foo(a, b)"))
         .equals("; EOF"));
 
-    //with more params and without whitespace after the parameter list
+    // with more params and without whitespace after the parameter list
     assert (serialize(p.parse(
         "#define foo(a, b...);\n"
             + "foo(a, b, c)"))

